### PR TITLE
Recent docs

### PR DESCRIPTION
**Pull request by cicithesquirrel**
Fix for #69.

What's to be done:
- Store the recent docs list in the config (max length of this list is in the config, too),
- Recent docs list gets updated when the user opens or saves a file,
- Add a "File > Recent" menu that shows the recent docs' paths (clicking an item will open the file),
- Add a "File > Recent > Clear" menu to clear the recent docs list.

### CODE REVIEW

**Comment by cicithesquirrel** [comment]
Current progress is:
- The first opened doc is not saved (but why?!)
- "File > Recent" menu is present but not filled/implemented

**Comment by brrd** [comment]
Hi,

Thanks again for your contribution.

I think you should have a look at [this](https://github.com/electron/electron/blob/master/docs/tutorial/desktop-environment-integration.md#recent-documents-windows--macos) for a better desktop integration on windows and osx.

In my opinion, recent docs history should not be stored in config.json for privacy reasons. Maybe not even in the config dir. What do you think about using [`localStorage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage) API instead? (so everything would be done in renderer process)

> The first opened doc is not saved (but why?!)

Maybe because of interferences between main and renderer processes (each of them are reading/writing in the config). Have you tried to call `storeRecentPath(this.path, this.id)` earlier? Like before `abrWin.open()` is called in `abrWin` constructor?

**Comment by cicithesquirrel** [comment]
I have troubles having a "recent files" menu that is dynamically filled and updated. This may be related to an Electron bug/limitation (https://github.com/electron/electron/issues/528)...

I found that I can:
- `myMenu.clear()` to destroy all of its items
- `myMenu.insert(0, new MenuItem({...}))` to insert an item 
But, the `clear` method is not documented in Electron API and I am worried I could create memory leaks using it :worried: 

I also tried to pre-create the menu items and then change their labels when the recent files list has changed. But it didn't work out: labels do not change.

File: app/abr-menu.js [hunk 1]
```diff
@@ -197,6 +199,50 @@ AbrMenu.prototype = {
     attach: function () {
         // FIXME: use win.setMenu(menu) for Linux and Windows instead
         Menu.setApplicationMenu(this.menu);
+    },
+
+    openRecentDoc: function (recentFile) {
+        this.abrWin.abrApp.open(recentFile);
+        // FIXCC recent-docs: Recent doc list is not updated after re-opening a recent file
```
Comment (cicithesquirrel) [inline app/abr-menu.js]: Remaining FIX/TODO: update recent doc list when reopening a recent file.

File: app/abr-menu.js [hunk 2]
```diff
@@ -197,6 +199,50 @@ AbrMenu.prototype = {
...
+            submenu.append(new MenuItem(itemData));
+        }
+
+        submenu.append(new MenuItem({
+            type: "separator"
+        }));
+
+        // append "clear recent" item
+        submenu.append(new MenuItem({
+            label: "Clear",
```
Comment (cicithesquirrel) [inline app/abr-menu.js]: Here, I could use the "i18n" branch (still to me merged "develop") to i18nize the "Clear" menu label.

**Comment by brrd** [comment]
I this PR ready for a review? I must confess that I lost track, all those updates confused me a bit 😕

**Comment by cicithesquirrel** [comment]
Yes, this PR is quite large...

In my opinion, there are two small things that could be fixed, even though they may be not essential:
* [A small bug fix (see comment in code)](https://github.com/brrd/Abricotine/pull/92#discussion-diff-78132274)
* [An i18n label](https://github.com/brrd/Abricotine/pull/92#discussion-diff-78132370) but merging i18n branch in develop is needed for this one.

But there is one thing that annoys me more: it is that I introduced a menu whose items change over time. This seems to go against [Electron possibilities](https://github.com/electron/electron/issues/528), and I was forced to use undocumented APIs (see my comment [above](https://github.com/brrd/Abricotine/pull/92#issuecomment-231857258)).

For fear of instability, one alternative would be to limit this feature [to simple OS integration](https://github.com/electron/electron/blob/master/docs/tutorial/desktop-environment-integration.md#recent-documents-windows--macos). As the doc says, it works on Windows (confirmed on Windows 10) and Mac (I cannot test this). On Ubuntu 16, there is no menu for selecting a recent doc from the taskbar, but the recent doc is listed in the global "recent files" menu, so it quite works...

So there is one decision to make: keep this PR "as is", or trash most of its lines of code to keep what is sure to work in the long term. That's a hard decision to make, and I may not be really objective ;-) So, tell me what you think!

**Comment by darahak** [comment]
VS Code has an ["Open Recent"](https://github.com/Microsoft/vscode/blob/1ba76bec9caa5735ac9c76140b39490bbd337ccb/src/vs/code/electron-main/menus.ts#L409) sub-menu which is similar to what you did.
It works fine, I don't see the problem.

I think this is a must-have, even though OS integration could be a nice addition.
FWIW, I ended up disabling it because I never used it (not intuitive for me).

**Comment by brrd** [comment]
First of all: thank you for this contribution.

Well, in my opinion we should only use OS integration and avoid the "open recent" menu for several reasons:

* As you pointed it out, dynamically updating menu items in electron is not easy (see [this](https://github.com/electron/electron/issues/527) and [this](https://github.com/electron/electron/issues/4028)). It's even more difficult when we must do it across windows. I don't say it's impossible, but it is tricky.
* We don't know if using `menuItem.clear()` is safe in term of performance and if it'll remain it electron API. I don't really like this...
* We still have to try your code on OSX and I'm not sure at all that it'll work because OSX menu bar behaves different in many ways.

Actually OS recent documents is easier to manage for us and it respects user's privacy settings. I personally don't like recent docs recording and I'm glad when I can switch it off with a single setting, which is the case in most OS.

So my idea would be to simply use OS integration right now and to wait for a better support of recent docs in Electron (or for some community module which would do the job in a reliable way) to create a dedicated menu.

But this is just my opinion, and as I told I don't use this feature. If you really think a menu would be useful then let's add it. I have to say that in such case we will have to work a bit more on it because I don't really agree with some parts of your code.

**Comment by cicithesquirrel** [comment]
In my point of view, a "recent docs" menu is an interesting feature: OS integration is usually not enough for me (but I wouldn't say it is a "must-have"). Your privacy concerns are understandable, and an option for enabling/disabling the menu could be useful. And the "clear" menu item can do the rest.

The option is also a way of avoiding any bugs on OSX, waiting for someone to test and contribute any patch that is needed.

As pointed by @darahak, we should look deeper into VSCode's way of doing this (in fact, I should have done this at first...).

That being said, if you still don't feel like merging this PR, it's OK with me. It won't hurt me or discourage my contributing to this project :-)

**Comment by brrd** [comment]
OK, it's useful so let's do it!

> As pointed by @darahak, we could look deeper into VSCode's way of doing this (in fact, I should have done this at first...).

Currently I see 2 ways to do this in electron:

1. Use `clear()` as you did.
2. Reprocess the whole menu bar each time the menu is updated (for each window). But this is a totally different approach and would better keep what you wrote for now in my opinion.

I need to review this PR first. I'm busy in this time so it'll wait a bit 😕 

Then we'll run tests on Windows/OSX/Linux. I got a VM for each OS now so I could maybe help.

**Comment by cicithesquirrel** [comment]
That's OK with me. Your "small" change requests have been fixed, I still need to work on the "storage". I'll keep you posted.

File: app/renderer/main.js [hunk 3]
```diff
@@ -7,6 +7,6 @@
 var AbrDocument = require.main.require("./abr-document.js");
 
 $( function () {
-    var abrDoc = new AbrDocument();
+    var abrDoc = new AbrDocument(localStorage);
```
Comment (brrd) [inline app/renderer/main.js]: Good! Don't forget this one.

Comment (cicithesquirrel) [inline app/renderer/main.js]: Sorry, I missed that one... Fixed.

**Comment by nettnikl** [comment]
Is this merge request still in work, or is it blocked some other issue?

**Comment by brrd** [comment]
@nettnikl This PR had no activity for almost 4 years and the related branch is 335 commits behind the current develop branch, so we can consider it obsolete unfortunately.

I'm closing this thread.

**Pull request closed without merging**